pennylane/vqe/vqe.py

```diff
@@ -370,70 +370,145 @@ class VQECost:
+        Grouping these commuting observables leads to fewer device executions:
 
-    >>> cost = qml.VQECost(ansatz, H, dev, interface="torch")
-    >>> params = torch.rand([4, 3])
-    >>> cost(params)
-    tensor(0.0245, dtype=torch.float64)
+        >>> cost_opt(params)
+        >>> ex_opt = dev.num_executions
+        >>> cost_no_opt(params)
+        >>> ex_no_opt = dev.num_executions - ex_opt
+        >>> print(ex_opt, ex_no_opt)
+        1 2
 
-    The cost function can be minimized using any gradient descent-based
-    :doc:`optimizer </introduction/optimizers>`.
+        Note that this feature is only available in :doc:`tape mode <../../code/qml_tape>`.
     """
 
     def __init__(
-        self, ansatz, hamiltonian, device, interface="autograd", diff_method="best", **kwargs
+        self,
+        ansatz,
+        hamiltonian,
+        device,
+        interface="autograd",
+        diff_method="best",
+        optimize=False,
+        **kwargs,
     ):
         coeffs, observables = hamiltonian.terms
+
         self.hamiltonian = hamiltonian
         """Hamiltonian: the hamiltonian defining the VQE problem."""
 
-        self.qnodes = qml.map(
-            ansatz, observables, device, interface=interface, diff_method=diff_method, **kwargs
-        )
-        """QNodeCollection: The QNodes to be evaluated. Each QNode corresponds to the
-        the expectation value of each observable term after applying the circuit ansatz.
-        """
+        self.qnodes = None
+        """QNodeCollection: The QNodes to be evaluated. Each QNode corresponds to the expectation
+        value of each observable term after applying the circuit ansatz."""
+
+        self._optimize = optimize
+
+        if self._optimize:
+            if not qml.tape_mode_active():
+                raise ValueError(
+                    "Observable optimization is only supported in tape mode. Tape "
+                    "mode can be enabled with the command:\n"
+                    "qml.enable_tape()"
+                )
+
+            obs_groupings, coeffs_groupings = qml.grouping.group_observables(observables, coeffs)
 
-        self.cost_fn = qml.dot(coeffs, self.qnodes)
+            wires = device.wires.tolist()
+
+            @qml.qnode(device, interface=interface, diff_method=diff_method, **kwargs)
+            def circuit(*qnode_args, obs, **qnode_kwargs):
+                """Converting ansatz into a full circuit including measurements"""
+                ansatz(*qnode_args, wires=wires, **qnode_kwargs)
+                return [qml.expval(o) for o in obs]
+
+            def cost_fn(*qnode_args, **qnode_kwargs):
+                """Combine results from grouped QNode executions with grouped coefficients"""
+                total = 0
+                for o, c in zip(obs_groupings, coeffs_groupings):
+                    res = circuit(*qnode_args, obs=o, **qnode_kwargs)
+                    total += sum([r * c_ for r, c_ in zip(res, c)])
```

Comment:
Written in such a way as to work on all interfaces 😄

Reply:
very nice!